### PR TITLE
Task 100 allure

### PR DESCRIPTION
**Pull request by ZagorskayaMarina**

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+            <scope>test</scope>
+        </dependency>
+        <dependency>
+            <groupId>org.junit.jupiter</groupId>
+            <artifactId>junit-jupiter-api</artifactId>
+            <version>5.7.0</version>
+            <scope>test</scope>
+        </dependency>
+        <dependency>
+            <groupId>io.qameta.allure</groupId>
```
Comment (artemmoroz) [inline pom.xml]: duplication (line 77)

File: src/test/java/driver/DriverSingleton.java [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+        }
+        return localInstance;
+    }
+
+    public WebDriver getDriver(Config config){
+        WebDriver driver = webDriver.get();
+        if (driver == null){
+            switch(config) {
+                case CHROME:
+                    ChromeOptions chromeOptions = new ChromeOptions();
```
Comment (artemmoroz) [inline src/test/java/driver/DriverSingleton.java:33]: you can try to use webdrivermanager lib to simplify dirver management https://github.com/bonigarcia/webdrivermanager

File: src/test/java/test_extensions/ScreenshotRules.java [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+import org.junit.jupiter.api.extension.TestWatcher;
+import org.openqa.selenium.Capabilities;
+import org.openqa.selenium.OutputType;
+import org.openqa.selenium.TakesScreenshot;
+import org.openqa.selenium.WebDriver;
+import org.openqa.selenium.remote.RemoteWebDriver;
+
+public class ScreenshotRules implements TestWatcher {
+
+    public WebDriver driver = DriverSingleton.getInstance().getDriver(Config.CHROME);
```
Comment (artemmoroz) [inline src/test/java/test_extensions/ScreenshotRules.java]: please remove Driver from here, cause you can have issues with multithreading. get it on demand. and why do you get always Chrome driver... what about FF)?

File: src/test/java/test_extensions/ScreenshotRules.java [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+
+    @Attachment(value = "Page screenshot", type = "image/png")
+    public byte[] saveScreenshot(byte[] screenShot) {
+        return screenShot;
+    }
+
+    public void takeScreenshot(){
+        WebDriver driver = DriverSingleton.getInstance().getDriver(Config.CHROME);
+        if (driver == null) {
+            System.out.println("");
```
Comment (artemmoroz) [inline src/test/java/test_extensions/ScreenshotRules.java]: very informative)

File: src/test/java/tests/yandex_email/BaseStep.java [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+import driver.Config;
+import driver.DriverSingleton;
+import org.junit.jupiter.api.AfterAll;
+import org.junit.jupiter.api.BeforeAll;
+import org.openqa.selenium.WebDriver;
+
+import static com.github.automatedowl.tools.AllureEnvironmentWriter.allureEnvironmentWriter;
+
+public class BaseStep {
+    static WebDriver driver;
```
Comment (artemmoroz) [inline src/test/java/tests/yandex_email/BaseStep.java]: bad idea to make this static, as methods below. Static is a very very bad word, and it brings a lot of issue in multithreading env. So, please read about static and thread, and remove it from here./

File: src/test/java/tests/yandex_email/BaseStep.java [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+
+    @BeforeAll
+    public static void initDriver(){
+        driver = DriverSingleton.getInstance().getDriver(Config.CHROME);
+    }
+    @BeforeAll
+    public static void setAllureEnvironment() {
+        allureEnvironmentWriter(
+                ImmutableMap.<String, String>builder()
+                        .put("Browser", "Chrome")
```
Comment (artemmoroz) [inline src/test/java/tests/yandex_email/BaseStep.java]: why do you set up these params explicitly? and what about FF

File: src/test/java/tests/yandex_email/YandexTests.java [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+    @Severity(SeverityLevel.CRITICAL)
+    @Description("User is authorized")
+    @Tag("stable")
+    public void verifyAuthorization(){
+        yandexHomePage = new YandexHomePage(driver);
+        try{
+            yandexHomePage.loginToSite();
+        }
+        catch (RuntimeException ex){
+            yandexHomePage.re_entry();
```
Comment (artemmoroz) [inline src/test/java/tests/yandex_email/YandexTests.java]: what next, if this step will fail?

File: src/test/java/tests/yandex_email/YandexTests.java [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+        }
+
+        Assertions.assertEquals(yandexHomePage.extractEmail(), "seleniumtests@tut.by", "User is authorized");
```
Comment (artemmoroz) [inline src/test/java/tests/yandex_email/YandexTests.java]: the expected result should be parametriezd. What if i will login with another email.

File: src/test/java/web_pages/YandexHomePage.java [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+        this.driver = driver;
+        PageFactory.initElements(this.driver, this);
+    }
+
+    @Step("Login to site")
+    public void loginToSite(){
+        WebDriverWait wait = new WebDriverWait(driver, 10);
+        driver.get("https://yandex.by/");
+        loginButton.click();
+        driver.findElement(By.id("passp-field-login")).sendKeys(properties.getProperty("USER_NAME"));
```
Comment (artemmoroz) [inline src/test/java/web_pages/YandexHomePage.java]: it is not a very cool approach, to use locators in the method body. Please select one of the widely used approaches (PageObject or PageFactory), cause in the current case your class is only called page but u did not receive any benefits from pageObject pattern.

File: src/test/java/web_pages/YandexHomePage.java [hunk 10]
```diff
@@ -0,0 +1,68 @@
...
+    public String extractEmail(){
+        WebDriverWait wait = new WebDriverWait(driver, 10);
+        return wait.until(ExpectedConditions.presenceOfElementLocated(By.xpath("//span[contains(@class, 'username ')]"))).getText();
+    }
+
+    @Step("Logout from site")
+    public void logOut(){
+        WebDriverWait wait = new WebDriverWait(driver, 10);
+        wait.until(ExpectedConditions.presenceOfElementLocated(By.xpath("//a/span[contains(@class, 'wrapper')]/span[contains(@class, 'avatar__image')]"))).click();
+        driver.findElement(By.xpath("//a/span[contains(@class, 'wrapper')]/span[contains(@class, 'avatar__image')]"));
```
Comment (artemmoroz) [inline src/test/java/web_pages/YandexHomePage.java]: why not to use css in that case, it seems to me that the locator will be simplier.

File: src/test/java/web_pages/YandexHomePage.java [hunk 11]
```diff
@@ -0,0 +1,68 @@
...
+import org.openqa.selenium.WebElement;
+import org.openqa.selenium.support.FindBy;
+import org.openqa.selenium.support.PageFactory;
+import org.openqa.selenium.support.ui.ExpectedConditions;
+import org.openqa.selenium.support.ui.WebDriverWait;
+import utils.Utils;
+
+import java.util.Properties;
+
+public class YandexHomePage {
```
Comment (artemmoroz) [inline src/test/java/web_pages/YandexHomePage.java:15]: please revisit this page class, and apply all patterns, approaches, parametrization, code duplication, etc. that we've already studied.

File: src/test/resources/mail.properties [hunk 12]
```diff
@@ -0,0 +1,5 @@
+USER_NAME_1=seleniumtests@tut.by
+PASSWORD_1=123456789zxcvbn
+
+USER_NAME=EE1VP@yandex.by
+PASSWORD=MarEE1!
```
Comment (artemmoroz) [inline src/test/resources/mail.properties:5]: I'm not able to add comment to screenshot files, please do not push them to repo, except you have a plan to do that.

**Review by artemmoroz — CHANGES_REQUESTED** [review]

File: pom.xml (hunk 1 above)
Comment (ZagorskayaMarina) [reply pom.xml]: fixed

File: src/test/java/test_extensions/ScreenshotRules.java (hunk 4 above)
Comment (ZagorskayaMarina) [reply src/test/java/test_extensions/ScreenshotRules.java]: fixed

File: src/test/java/tests/yandex_email/BaseStep.java (hunk 6 above)
Comment (ZagorskayaMarina) [reply src/test/java/tests/yandex_email/BaseStep.java]: fixed

File: src/test/java/web_pages/YandexHomePage.java (hunk 9 above)
Comment (ZagorskayaMarina) [reply src/test/java/web_pages/YandexHomePage.java]: fixed

File: src/test/java/tests/yandex_email/YandexTests.java (hunk 8 above)
Comment (ZagorskayaMarina) [reply src/test/java/tests/yandex_email/YandexTests.java]: fixed

File: src/test/java/test_extensions/ScreenshotRules.java (hunk 3 above)
Comment (ZagorskayaMarina) [reply src/test/java/test_extensions/ScreenshotRules.java]: fixed

File: src/test/java/tests/yandex_email/BaseStep.java (hunk 5 above)
Comment (ZagorskayaMarina) [reply src/test/java/tests/yandex_email/BaseStep.java]: fixed

File: src/test/java/tests/yandex_email/YandexTests.java (hunk 7 above)
Comment (ZagorskayaMarina) [reply src/test/java/tests/yandex_email/YandexTests.java]: fixed

File: src/test/java/web_pages/YandexHomePage.java (hunk 11 above)
Comment (ZagorskayaMarina) [reply src/test/java/web_pages/YandexHomePage.java:15]: fixed

**Pull request merged**